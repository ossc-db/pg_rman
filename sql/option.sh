#!/bin/bash

#============================================================================
# This is a test script for options of pg_rman.
#============================================================================

# Load common rules
. sql/common.sh option

# Check presence of pgbench command and initialize environment
which pgbench > /dev/null 2>&1
ERR_NUM=$?
if [ $ERR_NUM != 0 ]
then
    echo "pgbench is not installed in this environment."
    echo "It is needed in PATH for those regression tests."
    exit 1
fi

function cleanup()
{
	# cleanup environment
	pg_ctl stop -m immediate > /dev/null 2>&1
	rm -fr ${PGDATA_PATH}
	rm -fr ${BACKUP_PATH}
	rm -fr ${ARCLOG_PATH}
	rm -fr ${SRVLOG_PATH}
	rm -fr ${TBLSPC_PATH}
	mkdir -p ${ARCLOG_PATH}
	mkdir -p ${SRVLOG_PATH}
	mkdir -p ${TBLSPC_PATH}
}


function init_backup()
{
	# cleanup environment
	cleanup

	# create new database cluster
    initdb --no-locale -D ${PGDATA_PATH} > ${TEST_BASE}/initdb.log 2>&1
    cp ${PGDATA_PATH}/postgresql.conf ${PGDATA_PATH}/postgresql.conf_org
    cat << EOF >> ${PGDATA_PATH}/postgresql.conf
port = ${TEST_PGPORT}
logging_collector = on
wal_level = replica
log_directory = '${SRVLOG_PATH}'
archive_mode = on
archive_command = 'cp %p ${ARCLOG_PATH}/%f'
EOF

	# start PostgreSQL
	pg_ctl start -D ${PGDATA_PATH} -w -t 300 > /dev/null 2>&1
	pgbench -i -p ${TEST_PGPORT} -d postgres > ${TEST_BASE}/pgbench.log 2>&1

	# init backup catalog
	init_catalog
}

function init_catalog()
{
	rm -fr ${BACKUP_PATH}
	pg_rman init -B ${BACKUP_PATH} --quiet
}

cleanup

echo '###### COMMAND OPTION TEST-0001 ######'
echo '###### help option ######'
pg_rman --help;echo $?
echo ''

echo '###### COMMAND OPTION TEST-0002 ######'
echo '###### version option ######'
pg_rman --version;echo $?
echo ''

echo '###### COMMAND OPTION TEST-0003 ######'
echo '###### backup command failure without backup path option ######'
pg_rman backup -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0004 ######'
echo '###### backup command failure without arclog path option ######'
pg_rman backup -B ${BACKUP_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0005 ######'
echo '###### backup command failure without srvlog path option ######'
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -s -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0006 ######'
echo '###### backup command failure without backup mode option ######'
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0007 ######'
echo '###### backup command failure with invalid backup mode option ######'
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b bad -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0008 ######'
echo '###### delete failure without DATE ######'
pg_rman delete -B ${BACKUP_PATH};echo $?
echo ''

init_backup

echo '###### COMMAND OPTION TEST-0009 ######'
echo '###### syntax error in pg_rman.ini ######'
echo " = INFINITE" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0010 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "BACKUP_MODE=" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0011 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "COMPRESS_DATA=FOO" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0012 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_ARCLOG_FILES=TRUE" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0013 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_ARCLOG_DAYS=TRUE" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0014 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_SRVLOG_FILES=TRUE" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0015 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_SRVLOG_DAYS=TRUE" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0016 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_DATA_GENERATIONS=TRUE" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0017 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_DATA_GENERATIONS=0" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0018 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_SRVLOG_FILES=0" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0019 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "KEEP_SRVLOG_DAYS=0" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''


echo '###### COMMAND OPTION TEST-0020 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "SMOOTH_CHECKPOINT=FOO" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0021 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "WITH_SERVERLOG=FOO" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0022 ######'
echo '###### invalid value in pg_rman.ini ######'
init_catalog
echo "HARD_COPY=FOO" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0023 ######'
echo '###### invalid option in pg_rman.ini ######'
init_catalog
echo "TIMELINEID=1" >> ${BACKUP_PATH}/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -b full -p ${TEST_PGPORT};echo $?
echo ''

echo '###### COMMAND OPTION TEST-0024 ######'
echo '###### check priority of several pg_rman.ini files ######'
init_catalog
mkdir -p ${BACKUP_PATH}/conf_path_a
echo "BACKUP_MODE=ENV_PATH" > ${BACKUP_PATH}/pg_rman.ini
echo "BACKUP_MODE=ENV_PATH_A" > ${BACKUP_PATH}/conf_path_a/pg_rman.ini
pg_rman backup -B ${BACKUP_PATH} -A ${ARCLOG_PATH} -p ${TEST_PGPORT};echo $?
echo ''

# clean up the temporal test data
pg_ctl stop -m immediate > /dev/null 2>&1
rm -fr ${PGDATA_PATH}
rm -fr ${BACKUP_PATH}
rm -fr ${ARCLOG_PATH}
rm -fr ${SRVLOG_PATH}
rm -fr ${TBLSPC_PATH}
