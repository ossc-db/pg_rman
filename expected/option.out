\! bash sql/option.sh
###### COMMAND OPTION TEST-0001 ######
###### help option ######
pg_rman manage backup/recovery of PostgreSQL database.

Usage:
  pg_rman OPTION init
  pg_rman OPTION backup
  pg_rman OPTION restore
  pg_rman OPTION show [DATE]
  pg_rman OPTION show detail [DATE]
  pg_rman OPTION validate [DATE]
  pg_rman OPTION delete DATE
  pg_rman OPTION purge

Common Options:
  -D, --pgdata=PATH         location of the database storage area
  -A, --arclog-path=PATH    location of archive WAL storage area
  -S, --srvlog-path=PATH    location of server log storage area
  -B, --backup-path=PATH    location of the backup storage area
  -c, --check               show what would have been done
  -v, --verbose             show what detail messages
  -P, --progress            show progress of processed files

Backup options:
  -b, --backup-mode=MODE    full, incremental, or archive
  -s, --with-serverlog      also backup server log files
  -Z, --compress-data       compress data backup with zlib
  -C, --smooth-checkpoint   do smooth checkpoint before backup
  -F, --full-backup-on-error   switch to full backup mode
                               if pg_rman cannot find validate full backup
                               on current timeline
      NOTE: this option is only used in --backup-mode=incremental or archive.
  --keep-data-generations=NUM keep NUM generations of full data backup
  --keep-data-days=NUM        keep enough data backup to recover to N days ago
  --keep-arclog-files=NUM   keep NUM of archived WAL
  --keep-arclog-days=DAY    keep archived WAL modified in DAY days
  --keep-srvlog-files=NUM   keep NUM of serverlogs
  --keep-srvlog-days=DAY    keep serverlog modified in DAY days
  --standby-host=HOSTNAME   standby host when taking backup from standby
  --standby-port=PORT       standby port when taking backup from standby

Restore options:
  --recovery-target-time    time stamp up to which recovery will proceed
  --recovery-target-xid     transaction ID up to which recovery will proceed
  --recovery-target-inclusive whether we stop just after the recovery target
  --recovery-target-timeline  recovering into a particular timeline
  --hard-copy                 copying archivelog not symbolic link

Catalog options:
  -a, --show-all            show deleted backup too

Delete options:
  -f, --force               forcibly delete backup older than given DATE

Connection options:
  -d, --dbname=DBNAME       database to connect
  -h, --host=HOSTNAME       database server host or socket directory
  -p, --port=PORT           database server port
  -U, --username=USERNAME   user name to connect as
  -w, --no-password         never prompt for password
  -W, --password            force password prompt

Generic options:
  -q, --quiet               don't show any INFO or DEBUG messages
  --debug                   show DEBUG messages
  --help                    show this help, then exit
  --version                 output version information, then exit

Read the website for details. <http://github.com/ossc-db/pg_rman>
Report bugs to <http://github.com/ossc-db/pg_rman/issues>.
1

###### COMMAND OPTION TEST-0002 ######
###### version option ######
pg_rman 1.3.3
1

###### COMMAND OPTION TEST-0003 ######
###### backup command failure without backup path option ######
ERROR: required parameter not specified: BACKUP_PATH (-B, --backup-path)
12

###### COMMAND OPTION TEST-0004 ######
###### backup command failure without arclog path option ######
ERROR: required parameter not specified: ARCLOG_PATH (-A, --arclog-path)
12

###### COMMAND OPTION TEST-0005 ######
###### backup command failure without srvlog path option ######
ERROR: required parameter not specified: SRVLOG_PATH (-S, --srvlog-path)
12

###### COMMAND OPTION TEST-0006 ######
###### backup command failure without backup mode option ######
ERROR: required parameter not specified: BACKUP_MODE (-b, --backup-mode)
12

###### COMMAND OPTION TEST-0007 ######
###### backup command failure with invalid backup mode option ######
ERROR: invalid backup-mode "bad"
12

###### COMMAND OPTION TEST-0008 ######
###### delete failure without DATE ######
ERROR: delete range option not specified
HINT: Please run with 'pg_rman delete DATE'.
12

###### COMMAND OPTION TEST-0009 ######
###### syntax error in pg_rman.ini ######
WARNING: syntax error in " = INFINITE"
ERROR: required parameter not specified: BACKUP_MODE (-b, --backup-mode)
12

###### COMMAND OPTION TEST-0010 ######
###### invalid value in pg_rman.ini ######
ERROR: invalid backup-mode ""
12

###### COMMAND OPTION TEST-0011 ######
###### invalid value in pg_rman.ini ######
ERROR: option -Z, --compress-data should be a boolean: 'FOO'
12

###### COMMAND OPTION TEST-0012 ######
###### invalid value in pg_rman.ini ######
ERROR: option --keep-arclog-files should be a 32bit signed integer: 'TRUE'
12

###### COMMAND OPTION TEST-0013 ######
###### invalid value in pg_rman.ini ######
ERROR: option --keep-arclog-days should be a 32bit signed integer: 'TRUE'
12

###### COMMAND OPTION TEST-0014 ######
###### invalid value in pg_rman.ini ######
ERROR: option --keep-srvlog-files should be a 32bit signed integer: 'TRUE'
12

###### COMMAND OPTION TEST-0015 ######
###### invalid value in pg_rman.ini ######
ERROR: option --keep-srvlog-days should be a 32bit signed integer: 'TRUE'
12

###### COMMAND OPTION TEST-0016 ######
###### invalid value in pg_rman.ini ######
ERROR: option --keep-data-generations should be a 32bit signed integer: 'TRUE'
12

###### COMMAND OPTION TEST-0017 ######
###### invalid value in pg_rman.ini ######
ERROR: option -C, --smooth-checkpoint should be a boolean: 'FOO'
12

###### COMMAND OPTION TEST-0018 ######
###### invalid value in pg_rman.ini ######
ERROR: option -s, --with-serverlog should be a boolean: 'FOO'
12

###### COMMAND OPTION TEST-0019 ######
###### invalid value in pg_rman.ini ######
ERROR: option --hard-copy should be a boolean: 'FOO'
12

###### COMMAND OPTION TEST-0020 ######
###### invalid option in pg_rman.ini ######
ERROR: invalid option "TIMELINEID"
12

###### COMMAND OPTION TEST-0021 ######
###### check priority of several pg_rman.ini files ######
ERROR: invalid backup-mode "ENV_PATH"
12

